examples/Main.hs

```diff
@@ -289,3 +320,118 @@ dummyCurrency :: CurrencySymbol
 dummyCurrency =
   CurrencySymbol . fromJust . Aeson.decode $
     "\"1111111111111111111111111111111111111111111111111111111111111111\""
+
+shrinkerTests :: TestTree
+shrinkerTests =
+  testGroup
+    "shrinker tests"
```

Comment:
Preferably we want to avoid duplicating each test case. We probably need to restructure the examples. Do you have any opinions wrt. how to do this?

Reply:
Yeah I didn't feel great about that while writing it but couldn't think of a better way at the time. A data structure representing a test which could then be interpreted differently by the two sets of unit tests seems like a good way to go now that I think about it some more. If that sounds good to you I'll start working on that once I finish a bug fix in shrinker.

Reply:
That sounds fine to me! Another solution is making the `fails` etc. functions be members of a type class, such that you can instantiate them differently.

Reply:
A simple parametrization would work, I think?

```haskell
overallTests :: TestTree 
overallTests = 
  testGroup "tests" $
    flip fmap [CompileOpts True, CompileOpts False] $ plutarchTests

plutarchTests :: CompileOpts -> TestTree
plutarchTests opts =
  testGroup
    ("plutarch tests: shrink=" <> show (shrink opts))
       [ test "foo" $ do 
            compile opts (constant @PInteger 42) @?= 42
        ...
```

Instead of `compileAndShrink` - we could have `compile :: CompileOpts -> ...`, with:

```haskell
data CompileOpts = CompileOpts = { shrink :: Bool }
```

Tests that use `printTerm` can be separated out (since they are hardcoded to compare the UPLC). We can then write a `compile` that implicitly compares unshrunk and shrunk evaluation if shrink=True. With those two things in place, we wouldn't need free monads or type classes - and the test code would remain plain to use over time.

Reply:
That's a little tricky. For the PrintsTo tests I don't just compile it with shrink and make sure it prints to the target string (because that would only work when shrink doesn't change anything) I just check that the output with and without shrink are the same (up to a loose notion of equivalence from shrinker-testing). Also with Equal I also add some extra equality tests between evaluations with and without shrink. 

I think all the other tests could be made more generic along those lines by replacing them all with `Test :: (Reader CompileOpts Assertion) -> TestStruct` but I think I'd still need the free monad for PrintsTo and Equal.